### PR TITLE
Invalidate all

### PR DESCRIPTION
**Pull request by redwarewolf**
## Summary

- /invalidate_all endpoint

### CODE REVIEW

File: app/services/users.js [hunk 1]
```diff
@@ -46,3 +46,19 @@ exports.index = async req => {
...
+
+const findUserByMail = mail => User.findOne({ where: { mail } });
+
+const invalidateAll = token => {
+  try {
+    return findUserByMail(jwt.validate(token).mail).then(user => {
+      user.tokenEmitDate = Date.now();
+      user.save();
+      return user;
+    });
```
Comment (picollomartin) [inline app/services/users.js:58]: Avoid mixing services, also you only need an update query for this and you're doing two queries here

File: test/controllers/user.test.js [hunk 2]
```diff
@@ -213,4 +214,31 @@ describe('Users Controller', () => {
...
+  describe('/POST users/sessions/invalidate_all', () => {
+    describe('when using valid parameters', () => {
+      it('should invalidate all session tokens', done => {
+        factory.create('User', { password: 'QSShBtjP' }).then(user => {
+          request
+            .post('/api/v1/users/sessions')
+            .send({
+              mail: user.mail,
+              password: '$2b$10$G4b27Ilbv5Jmg8IrtWzjUuiY3zD2wvG9OuWlXEbg60F5Xy8s1Z12u'
+            })
```
Comment (picollomartin) [inline test/controllers/user.test.js:227]: Avoid test two endpoints together, generate fake data directly (valid tokens) and test the endpoint/middleware

File: test/controllers/user.test.js [hunk 3]
```diff
@@ -213,4 +214,31 @@ describe('Users Controller', () => {
...
+            .set('Accept', 'application/json')
+            .then(res => {
+              request
+                .post('/api/v1/users/sessions/invalidate_all')
+                .set({ Accept: 'application/json', authorization: res.headers.authorization })
+                .then(() => {
+                  expect(
+                    jwt.validate(res.headers.authorization).iat <= Math.round(user.tokenEmitDate / 1000)
+                  ).toBeTruthy();
```
Comment (picollomartin) [inline test/controllers/user.test.js:236]: You should validate some success case also

File: app/middlewares/authorizationValidator.js [hunk 4]
```diff
@@ -1,10 +1,15 @@
...
 
-exports.validate = (req, res, next) => {
+exports.validate = async (req, res, next) => {
   try {
-    jwt.validate(req.headers.authorization);
-    next();
+    const token = jwt.validate(req.headers.authorization);
+    const user = await userService.findUserByMail(token.mail);
+    if (user.tokenEmitDate / 1000 <= token.iat) {
+      next();
```
Comment (picollomartin) [inline app/middlewares/authorizationValidator.js:10]: Add return here !

File: app/controllers/users.js [hunk 5]
```diff
@@ -32,6 +32,13 @@ exports.getUsers = (req, res, next) => {
     .catch(error => next(error));
 };
 
+exports.invalidateAll = (req, res, next) => {
+  usersService
+    .invalidateAll(jwt.validate(req.headers.authorization))
```
Comment (picollomartin) [inline app/controllers/users.js:37]: Why do you need to validate again the token ?

Comment (redwarewolf) [reply app/controllers/users.js:37]: It's actually a decode, the name just stuck from older functionalities but it doesn't fit anymore. I'll change it

Comment (picollomartin) [reply app/controllers/users.js:37]: You must decode and validate in the middleware, not in each controller

File: app/middlewares/validators/adminValidator.js [hunk 6]
```diff
@@ -1,10 +1,17 @@
 const errors = require('../../errors');
 const jwt = require('../../services/jwt');
+const userService = require('../../services/users');
 
-exports.validateAdmin = (req, res, next) => {
+exports.validate = async (req, res, next) => {
```
Comment (picollomartin) [inline app/middlewares/validators/adminValidator.js:5]: this is a validateAdmin

**Pull request merged**